MetaMask/metamask-extension
# Allow customizing order of bify transforms/plugins

We use Browserify to compile our JavaScript code into a distributable
format. You can use Browserify in a few different ways, but we use the
API, which is detailed [here][1]. The way we use it is as follows:

``` javascript
const bundlerOpts = {
  entries: [],
  transform: [],
  plugin: [],
  require: [],
  manualExternal: [],
  manualIgnore: [],
}
// ... change bundlerOpts in various ways ...

// then, later:
const bundler = browserify(bundlerOpts);
bundler.bundle();
```

Note that all transforms and all plugins are passed to `browserify` in
one go. This creates a problem, because transforms and plugins are
processed in the same order in which they are specified. That is,
specifying transform A, then plugin B, then transform C, is NOT the same
as specifying transform A, then transform C, then plugin B.

The way to correct this is to make use of Browserify's full API, which
provides `transform` and `plugin` methods, as well as other specific
methods to specify other options. This commit, then, updates the part of
the build system that configures Browserify to use this API.

The order dependency as it relates to transforms and plugins will come
into play when introducing TypeScript into the build system. Namely, we
will be introducing a `tsify` Browserify plugin, which will need to be
specified before specifying `babelify`, as demonstrated [here][2]. Hence
we need this commit to be in place so that this is possible.

[1]: https://www.npmjs.com/package/browserify#browserifyfiles--opts
[2]: https://github.com/TypeStrong/tsify#es2015-formerly-known-as-es6

---

Manual testing steps:  

* Make sure you've added the development version of the extension to your browser (sourced from `dist/`).
* Run `yarn start`. Reload the extension if need be. You should not see a white screen. Click around in the extension for a bit. You should not encounter any error screens.
* Run `yarn start:test`. Run an e2e test. It should pass.
* Run `yarn dist`. Reload the extension if need be. You should not see a white screen. Click around in the extension for a bit. You should not encounter any error screens.
